### PR TITLE
PUP-(8598, 8622, 8635, 8640, 8641, 8657, 8676) Add Feature Deferred Type and related functionality

### PR DESCRIPTION
**Pull request by hlindberg**
This adds the Deferred data type and related functionality (needed or wanted when using Deferred values).

* Makes it possible to call functions (in core puppet) on the agent
* Makes it possible to use a Deferred data type that is resolved by making a function call on the agent side before catalog is used
* Adds (missing ability) concatenation of two Binary values.
* Makes it possible to resolve a Deferred by calling it (for debugging purposes and if used as a means to do late binding call
  while compiling a manifest).
* Handles Sensitive flag depending on outcome of Deferred resolution (agent side)
* Adds a cache to the function API in anticipation that deferred functions typically do communication and want to cache
  a connection or token of some kind.

Some tests are updated as (maint) commits - that may have been required because I ran on newer Ruby version, but the changes are not harmful and works on older Ruby's as well.

Also see #6899 (Encrypted data type) which contains the same commits as this PR.

Note that this is for 5.5.x as an experimental feature - when discussed earlier it was decided that work should be done against 5.5.x to make it possible to release in the 5 series if we wanted. That may no longer be true, and if so everything needs a rebase to 6.0.

### CODE REVIEW

File: lib/puppet/application/apply.rb [hunk 1]
```diff
@@ -265,20 +265,31 @@ def main
...
 
-        exit_status = Puppet.override(:loaders => Puppet::Pops::Loaders.new(apply_environment)) { apply_catalog(catalog) }
+          catalog.retrieval_duration = Time.now - starttime
 
+          if options[:write_catalog_summary]
+            catalog.write_class_file
+            catalog.write_resource_file
+          end
+
+          #exit_status = Puppet.override(:loaders => Puppet::Pops::Loaders.new(apply_environment)) { apply_catalog(catalog) }
```
Comment (pcarlisle) [inline lib/puppet/application/apply.rb]: don't need to keep this

File: lib/puppet/configurer.rb [hunk 2]
```diff
@@ -4,6 +4,8 @@
 require 'puppet/network/http_pool'
 require 'puppet/util'
 require 'securerandom'
+#require 'puppet/parser/script_compiler'
```
Comment (pcarlisle) [inline lib/puppet/configurer.rb]: ?

File: lib/puppet/configurer/fact_handler.rb [hunk 3]
```diff
@@ -28,7 +28,11 @@ def find_facts
   end
 
   def facts_for_uploading
-    facts = find_facts
+    encode_facts(find_facts)
+  end
+
+  def encode_facts(facts)
+    #facts = find_facts
```
Comment (pcarlisle) [inline lib/puppet/configurer/fact_handler.rb]: don't need

File: lib/puppet/pops/evaluator/deferred_resolver.rb [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+      result
+    elsif x.is_a?(Puppet::Pops::Types::PSensitiveType::Sensitive)
+      # rewrap in a new Sensitive after resolving any nested deferred values
+      Puppet::Pops::Types::PSensitiveType::Sensitive.new(resolve(x.unwrap))
+    else
+      x
+    end
+  end
+
+  def resolve_future(f)
```
Comment (thallgren) [inline lib/puppet/pops/evaluator/deferred_resolver.rb]: I react to the term "future" here. It's a somewhat loaded word that doesn't really describe a deferred value. There's no blocking wait for something to happen involved here and no description of something to wait for either. So, `resolve_deferred` perhaps (same in plural form i name of caller function)?

**Comment by tvpartytonight** [comment]
> Makes it possible to call functions (in core puppet) on the agent

Is there a reason why this is limited to only core functions? What would it take for this to call the autoloader to load in custom functions defined in modules?

**Comment by joshcooper** [comment]
Superseded by https://github.com/puppetlabs/puppet/pull/6951

**Pull request closed without merging**